tfhe/src/shortint/ciphertext/compressed_modulus_switched_ciphertext.rs

```diff
@@ -0,0 +1,155 @@
+
+/// An object to store a ciphertext in little memory.
+/// Decompressing it requires a PBS
+///
+/// # Example
+///
+/// ```rust
+/// use tfhe::shortint::gen_keys;
+/// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+///
+/// // Generate the client key and the server key:
+/// let (cks, sks) = gen_keys(PARAM_MESSAGE_2_CARRY_2_KS_PBS);
+///
+/// let clear = 3;
+///
+/// let ctxt = cks.unchecked_encrypt(clear);
+///
+/// // Can be serialized in a smaller buffer
+/// let compressed_ct = sks.switch_modulus_and_compress(&ctxt);
+///
+/// let decompressed_ct = sks.decompress(&compressed_ct);
+///
+/// let dec = cks.decrypt(&decompressed_ct);
+///
+/// assert_eq!(clear, dec);
+/// ```
+#[derive(Clone, serde::Serialize, serde::Deserialize)]
+pub struct CompressedModulusSwitchedCiphertext {
+    compressed_modulus_switched_lwe_ciphertext: CompressedModulusSwitchedLweCiphertext<u64>,
+    degree: Degree,
+    message_modulus: MessageModulus,
+    carry_modulus: CarryModulus,
+    pbs_order: PBSOrder,
+}
+
+impl ServerKey {
+    // Packs a ciphertext to have a smaller serialization size
+    pub fn switch_modulus_and_compress(
+        &self,
+        ct: &Ciphertext,
+    ) -> CompressedModulusSwitchedCiphertext {
+        let compressed_modulus_switched_lwe_ciphertext = match self.pbs_order {
+            PBSOrder::KeyswitchBootstrap => {
+                let ms_ed = ShortintEngine::with_thread_local_mut(|engine| {
+                    let (mut ciphertext_buffers, _) = engine.get_buffers(self);
+
+                    keyswitch_lwe_ciphertext(
+                        &self.key_switching_key,
+                        &ct.ct,
+                        &mut ciphertext_buffers.buffer_lwe_after_ks,
+                    );
+
+                    CompressedModulusSwitchedLweCiphertext::compress(
+                        &ciphertext_buffers.buffer_lwe_after_ks,
+                        CiphertextModulusLog(self.bootstrapping_key.polynomial_size().log2().0 + 1),
+                    )
+                });
+
+                ms_ed
+            }
+            PBSOrder::BootstrapKeyswitch => CompressedModulusSwitchedLweCiphertext::compress(
+                &ct.ct,
+                CiphertextModulusLog(self.bootstrapping_key.polynomial_size().log2().0 + 1),
+            ),
+        };
+
+        CompressedModulusSwitchedCiphertext {
+            compressed_modulus_switched_lwe_ciphertext,
+            degree: ct.degree,
+            message_modulus: ct.message_modulus,
+            carry_modulus: ct.carry_modulus,
+            pbs_order: ct.pbs_order,
+        }
+    }
+
+    // Unpacks a compressed ciphertext
+    // The degree from before the packing is conserved.
+    // This operation uses a PBS. Fot the same cost, it's possible to apply a lookup table by
+    // calling unpack_and_apply_lookup_table instead.
+    pub fn decompress(&self, compressed_ct: &CompressedModulusSwitchedCiphertext) -> Ciphertext {
```

Comment:
docstring thing, especially since you already have it written out basically :)

Reply:
Done